### ISSUE TITLE
Global starting point for factory.Sequence counters

### ISSUE THREAD

**Issue by gpip** [rbarrois/factory_boy#283]
After reading the documentation it seems there's no supported way of defining a starting point for all factories. I'm looking for something similar to:

```
factory.COUNTER_START = n
obj1 = SomeFactory()
obj2 = SomeFactory()
obj3 = SomeOtherFactory()
assert(obj1.someattr == factory.COUNTER_START)
assert(obj2.someattr == factory.COUNTER_START + 1)
assert(obj3.someattr == factory.COUNTER_START)
```

When running the tests it helps in having an unique COUNTER_START per process so you can run similar tests in multiple processes without having database conflicts; also if you want to preserve past test data and re-execute tests.

Right now it seems all the classes need to be modified to take something like COUNTER_START into account, so this issue is about having a helper that does that for all the factories (maybe a factory could optionally want to disable the feature though?).

**Comment by gpip** [comment]
It causes conflicts on unique attributes, maybe I should be using a different method for picking those values (but if that's the case, aren't we going into a loop into this being easy vs being doable)? For now I'm using isolated databases for tests running in multiple processes.

I guess we're thinking about different cases, because the data sent via FactoryBoy is certainly not the same the application sends. If conflicting information is sent via the app, I want that to fail (be either at the db layer, or above it). In FactoryBoy's case, I want to avoid sending conflicting information unless I'm testing for that.

**Comment by jeffwidman** [comment]
> I guess we're thinking about different cases, because the data sent via FactoryBoy is certainly not the same the application sends. 

You're completely right, I do this myself. Somehow my brain started assuming that you were hardcoding PK IDs (even though you didn't say that) which raises red flags for me. When I first switched to running all my tests within separate DB transactions so that I could roll them back rather than having to drop/recreate the whole DB for every test, a bunch of tests started failing. Took me forever to track down the issue. I had asserts testing the PK IDs were what I expected, but when you roll back a transaction, the auto-increment of the PK doesn't get reset (at least in PostgreSQL). Anyway, sorry about the confusion.

> It causes conflicts on unique attributes, maybe I should be using a different method for picking those values (but if that's the case, aren't we going into a loop into this being easy vs being doable)? For now I'm using isolated databases for tests running in multiple processes.

Can you describe the columns/unique constraints a little more? 

I'm still not convinced that the right solution here is having a global counter because I've personally experienced similar issues where the easier solution was something else. For example, I hit unique constraint violations using `faker.words()` where I had more values to insert than there were words in the pool. I solved it by tacking on an incrementing integer at the end of the word.

**Comment by gpip** [comment]
> Can you describe the columns/unique constraints a little more?

Let's simplify the situation and assume there's one database which does not get dropped between test runs. So this database X has a table U, and U has an unique field named username.

Using FactoryBoy I create a class attribute to define an username exactly like this one: `username = factory.Sequence(lambda n: 'user%d' %d)`. It seems to be doing it in a similar way to what you mentioned, tackling an integer at the end. Now, I could go further on this and do exactly like you: pick a word at random from a dictionary instead of prefixing with "user". But if I run tests that use this many times (remember, in this example the database is never dropped), I still have a good chance to hit a conflicting username.

But, what's `n` for FactoryBoy? What if `n` was the current timestamp? If `n` was the current timestamp then I could re-run the same tests forever without conflicting on usernames (assuming the machine(s) is/are running ntp, etc).

In the situation where there is a single database which gets dropped between test runs but multiple processes hitting it, `n` would start at 0 for each process and the tests would cause expected integrity errors. If `n` started with (just making a guess here) the `PID || timestamp` then, again, the tests wouldn't run into conflicts. The exact process of defining a starting point depends on how you manage test runs, if you can pass an integer (starting from `0`) to processes/threads you're spawing then prepend that to the timestamp as a string to avoid the integrity errors.

All of this does not really matter if you're using `x` processes and each one uses its own copy of database `X`, since you can start all counters from 0 and they will not run into the problem.

**Comment by rbarrois** [comment]
@gpip Even if this does not match typical factory_boy use cases, I see how that could be useful, mostly for the "multiple parallel runs of tests in threads".

So, if you're ready to provide the code, I'd be glad to help you along ;)

A good starting point would be ``BaseFactory._setup_next_sequence()`` in ``factory/base.py``.
